src/sqerl_core.erl

```diff
@@ -0,0 +1,260 @@
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+-module(sqerl_core).
+
+-include("sqerl.hrl").
+-define(MAX_RETRIES, 5).
+
+%% See http://www.postgresql.org/docs/current/static/errcodes-appendix.html
+-define(PGSQL_ERROR_CODES, [{<<"23505">>, conflict}, {<<"23503">>, foreign_key}]).
+
+-export([checkout/1,
+         checkin/2,
+         with_db/2,
+         with_db/3,
+         execute/3,
+         execute_statement/5,
+         bulk_insert/6,
+         parse_error/1,
+         parse_statement_results/1,
+         parse_select_results/1,
+         extract_insert_data/1]).
+
+checkout(Pool) ->
+    pooler:take_member(Pool, envy:get(sqerl, pooler_timeout, 0, integer)).
+
+checkin(Pool, Connection) ->
+    pooler:return_member(Pool, Connection).
+
+with_db(Pool, Call) ->
+    with_db(Pool, Call, ?MAX_RETRIES).
+
+with_db(_Pool, _Call, 0) ->
+    {error, no_connections};
+with_db(Pool, Call, Retries) ->
+    case checkout(Pool) of
+        error_no_members ->
+            {error, no_connections};
+        Cn when is_pid(Cn) ->
+            %% We don't need a try/catch around Call(Cn) because pooler links both the
+            %% connection and the process that has the connection checked out (this
+            %% process). So a crash here will not leak a connection.
+            case Call(Cn) of
+                {error, closed} ->
+                    %% Closing the connection will cause the process
+                    %% to shutdown. pooler will get notified and
+                    %% remove the connection from the pool.
+                    sqerl_client:close(Cn),
+                    with_db(Call, Retries - 1);
+                Result ->
+                    checkin(Pool, Cn),
+                    Result
+            end
+    end.
+
+execute_statement(Pool, StmtName, StmtArgs, XformName, XformArgs) ->
+    case execute(Pool, StmtName, StmtArgs) of
+        {ok, Results} ->
+            Xformer = erlang:apply(sqerl_transformers, XformName, XformArgs),
+            Xformer(Results);
+        {ok, Count, Results} ->
+            %% we'll get here for an INSERT ... RETURNING query
+            Xformer = erlang:apply(sqerl_transformers, XformName, XformArgs),
+            {ok, XResult} = Xformer(Results),
+            {ok, Count, XResult};
+        Other ->
+            Other
+    end.
+
+%% @doc Execute query or statement with parameters.
+%% ```
+%% Returns:
+%% - {ok, Result}
+%% - {error, ErrorInfo}
+%%
+%% Result depends on the query being executed, and can be
+%% - Rows
+%% - Count
+%%
+%% Row is a proplist-like array, e.g. [{<<"id">>, 1}, {<<"name">>, <<"John">>}]
+%%
+%% Note that both a simple query and a prepared statement can take
+%% parameters.
+%% '''
+%%
+-spec execute(atom(), sqerl_query(), [] | [term()]) -> sqerl_results().
+execute(Pool, QueryOrStatement, Parameters) ->
```

Comment:
One thing I'm wondering about is that this also provides a means of specifying the connection. That might help us add transaction support as well as use a different pattern of connection use. For example, you could pass `{conn, Pid}` instead of the the pool atom.